internal/lib/checkpoint.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/containers/podman/v3/pkg/annotations"
+	"github.com/containers/podman/v3/pkg/checkpoint/crutils"
+	"github.com/containers/storage/pkg/archive"
+	"github.com/cri-o/cri-o/internal/oci"
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/generate"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type ContainerCheckpointRestoreOptions struct {
+	Container string
+	Pod       string
+
+	libpod.ContainerCheckpointOptions
+}
+
+// ContainerCheckpoint checkpoints a running container.
+func (c *ContainerServer) ContainerCheckpoint(ctx context.Context, opts *ContainerCheckpointRestoreOptions) (string, error) {
+	ctr, err := c.LookupContainer(opts.Container)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to find container %s", opts.Container)
+	}
+
+	configFile := filepath.Join(ctr.BundlePath(), "config.json")
+	specgen, err := generate.NewFromFile(configFile)
+	if err != nil {
+		return "", errors.Wrapf(err, "Not able to read config for container %q", ctr.ID())
+	}
+
+	cStatus := ctr.State()
+	if cStatus.Status != oci.ContainerStateRunning {
+		return "", fmt.Errorf("container %s is not running", ctr.ID())
+	}
+
+	if opts.TargetFile != "" {
+		if err := c.prepareCheckpointExport(ctr); err != nil {
+			return "", errors.Wrapf(err, "failed to write config dumps for container %s", ctr.ID())
+		}
+	}
+
+	if err := c.runtime.CheckpointContainer(ctx, ctr, specgen.Config, opts.KeepRunning); err != nil {
+		return "", errors.Wrapf(err, "failed to checkpoint container %s", ctr.ID())
+	}
+	if opts.TargetFile != "" {
+		if err := c.exportCheckpoint(ctr, specgen.Config, opts.TargetFile); err != nil {
+			return "", errors.Wrapf(err, "failed to write file system changes of container %s", ctr.ID())
+		}
+	}
+	if err := c.storageRuntimeServer.StopContainer(ctr.ID()); err != nil {
+		return "", errors.Wrapf(err, "failed to unmount container %s", ctr.ID())
+	}
+	if err := c.ContainerStateToDisk(ctx, ctr); err != nil {
+		logrus.Warnf("Unable to write containers %s state to disk: %v", ctr.ID(), err)
+	}
+
+	if !opts.Keep {
+		cleanup := []string{
+			"dump.log",
+			"stats-dump",
+			metadata.ConfigDumpFile,
+			metadata.SpecDumpFile,
+		}
+		for _, del := range cleanup {
+			file := filepath.Join(ctr.Dir(), del)
+			if err := os.Remove(file); err != nil {
+				logrus.Debugf("Unable to remove file %s", file)
+			}
+		}
+	}
+
+	return ctr.ID(), nil
+}
+
+// Copied from libpod/diff.go
+var containerMounts = map[string]bool{
+	"/dev":               true,
+	"/etc/hostname":      true,
+	"/etc/hosts":         true,
+	"/etc/resolv.conf":   true,
+	"/proc":              true,
+	"/run":               true,
+	"/run/.containerenv": true,
+	"/run/secrets":       true,
+	"/sys":               true,
+}
+
+// getDiff returns the file system differences
+// Copied from libpod/diff.go and simplified for the checkpoint use case
+func (c *ContainerServer) getDiff(id string) ([]archive.Change, error) {
```

Comment:
nit:
```suggestion
func (c *ContainerServer) getDiff(id string) (rchanges []archive.Change, err error) {
```
Then we can remove line 106

Reply:
changed